### PR TITLE
[RFC] getcompletion() can get cmdline-completion result

### PR DESCRIPTION
**Pull request by Shougo**
It is the port of https://github.com/vim/vim/pull/1140.
It is already todo listed in Vim.

### CODE REVIEW

**Comment by Shougo** [comment]
@justinmk Can you check the patch?
It is useful for the plugins.

For example, neco-vim can generate the completion result more generally.
You can use denite.nvim as command line completion.

**Comment by thalesmello** [comment]
:+1:

File: src/nvim/eval.c [hunk 1]
```diff
@@ -10129,28 +10129,38 @@ static void f_getcompletion(typval_T *argvars, typval_T *rettv, FunPtr fptr)
...
-    xpc.xp_pattern_len = STRLEN(xpc.xp_pattern);
-  }
+      if (xpc.xp_context == EXPAND_NOTHING) {
+          if (argvars[1].v_type == VAR_STRING) {
+              EMSG2(_(e_invarg2), argvars[1].vval.v_string);
+          } else {
+              EMSG(_(e_invarg));
+          }
+          return;
+      }
```
Comment (justinmk) [inline src/nvim/eval.c]: indentation

File: runtime/doc/eval.txt [hunk 2]
```diff
@@ -3812,14 +3813,19 @@ getcompletion({pat}, {type} [, {filtered}])		*getcompletion()*
 		user		user names
 		var		user variables
 
-		If {pat} is an empty string, then all the matches are returned.
-		Otherwise only items matching {pat} are returned. See
-		|wildcards| for the use of special characters in {pat}.
+		If {pat} is an empty string, then all the matches are
+		returned.  Otherwise only items matching {pat} are returned.
```
Comment (justinmk) [inline runtime/doc/eval.txt]: why was this newline added?

File: src/nvim/eval.c [hunk 3]
```diff
@@ -10129,28 +10129,38 @@ static void f_getcompletion(typval_T *argvars, typval_T *rettv, FunPtr fptr)
...
   xpc.xp_context = cmdcomplete_str_to_type(get_tv_string(&argvars[1]));
-  if (xpc.xp_context == EXPAND_NOTHING) {
-    if (argvars[1].v_type == VAR_STRING) {
-      EMSG2(_(e_invarg2), argvars[1].vval.v_string);
-    } else {
-      EMSG(_(e_invarg));
-    }
-    return;
-  }
+  if (strcmp((char *)get_tv_string(&argvars[1]), "cmdline") == 0) {
```
Comment (justinmk) [inline src/nvim/eval.c]: here `argvars[1]` is used  before `argvars[1].v_type == VAR_STRING` validates it.

File: src/nvim/testdir/test_cmdline.vim [hunk 4]
```diff
@@ -151,6 +151,20 @@ func Test_getcompletion()
...
+  let l = getcompletion('cd NoMatch', 'cmdline')
+  call assert_equal([], l)
+  let l = getcompletion('let v:n', 'cmdline')
+  call assert_true(index(l, 'v:null') >= 0)
+  let l = getcompletion('let v:notexists', 'cmdline')
+  call assert_equal([], l)
+  let l = getcompletion('call tag', 'cmdline')
+  call assert_true(index(l, 'taglist(') >= 0)
+  let l = getcompletion('call paint', 'cmdline')
+  call assert_equal([], l)
```
Comment (justinmk) [inline src/nvim/testdir/test_cmdline.vim]: There are more kinds of completion, would be good to test more kinds. 

Also, patches sent to Neovim, which are not part of Vim, should be tested in `test/functional/`. `test/functional/viml/completion_spec.lua` exists, can just create a new test there.

File: runtime/doc/eval.txt (hunk 2 above)
Comment (Shougo) [inline runtime/doc/eval.txt]: Because, the line overs 78 column.
I just have formatted.

File: src/nvim/eval.c (hunk 3 above)
Comment (Shougo) [inline src/nvim/eval.c]: Oh.  OK.
I will restore the lines.

File: src/nvim/testdir/test_cmdline.vim (hunk 4 above)
Comment (Shougo) [inline src/nvim/testdir/test_cmdline.vim]: > There are more kinds of completion, would be good to test more kinds.

Well, if possible....  Vim/neovim has kinds of command line completion types.
It is hard to cover all kinds test.

This feature is not for neovim only.
I don't want to convert the tests.

**Comment by Shougo** [comment]
Thanks.  I have fixed.

**Comment by justinmk** [comment]
Still have the same question mentioned in https://github.com/neovim/neovim/pull/5823#discussion_r94514937 .

**Comment by Shougo** [comment]
> Still have the same question mentioned in #5823 (comment) .

I have answered to it.

**Comment by Shougo** [comment]
I have fixed the lint error

File: src/nvim/eval.c [hunk 5]
```diff
@@ -10291,19 +10291,37 @@ static void f_getcompletion(typval_T *argvars, typval_T *rettv, FunPtr fptr)
     return;
   }
 
-  if (xpc.xp_context == EXPAND_MENUS) {
-    set_context_in_menu_cmd(&xpc, (char_u *)"menu", xpc.xp_pattern, false);
-    xpc.xp_pattern_len = STRLEN(xpc.xp_pattern);
-  }
+  if (strcmp((char *)get_tv_string(&argvars[1]), "cmdline") == 0) {
```
Comment (justinmk) [inline src/nvim/eval.c]: This is still reference `argvars[1]` before it is validated (line 10303).

**Comment by Shougo** [comment]
It does not work...
I will fix it.

**Comment by Shougo** [comment]
I have fixed the errors.

**Comment by ezuk** [comment]
Is this getting merged in?

**Comment by Shougo** [comment]
Still pending?

**Comment by raichoo** [comment]
Maybe mark it as [RDY] to attract some attention ^^?

**Comment by Shougo** [comment]
I have fixed the conflict.

> Wondering if I could use this to pipe completions results into FZF?

I think you can use it in fzf.vim.

**Comment by Shougo** [comment]
The tests are passed.
I think it is [RDY].

File: src/nvim/eval.c [hunk 6]
```diff
@@ -10764,32 +10764,40 @@ static void f_getcompletion(typval_T *argvars, typval_T *rettv, FunPtr fptr)
...
+    return;
+  }
+
   ExpandInit(&xpc);
   xpc.xp_pattern = get_tv_string(&argvars[0]);
-  xpc.xp_pattern_len = (int)STRLEN(xpc.xp_pattern);
-  xpc.xp_context = cmdcomplete_str_to_type(get_tv_string(&argvars[1]));
-  if (xpc.xp_context == EXPAND_NOTHING) {
-    if (argvars[1].v_type == VAR_STRING) {
+  xpc.xp_pattern_len = STRLEN(xpc.xp_pattern);
```
Comment (justinmk) [inline src/nvim/eval.c]: this is immediately overwritten by the if{} block. Can it be moved to the else{} block?

File: src/nvim/testdir/test_cmdline.vim [hunk 7]
```diff
@@ -156,6 +156,20 @@ func Test_getcompletion()
     call assert_equal(['Testing'], l)
   endif
 
+  " Command line completion tests
+  let l = getcompletion('cd ', 'cmdline')
+  call assert_true(index(l, 'sautest/') >= 0)
+  let l = getcompletion('cd NoMatch', 'cmdline')
+  call assert_equal([], l)
+  let l = getcompletion('let v:n', 'cmdline')
+  call assert_true(index(l, 'v:null') >= 0)
```
Comment (justinmk) [inline src/nvim/testdir/test_cmdline.vim]: why not use `assert_match` ?

File: src/nvim/eval.c [hunk 8]
```diff
@@ -10764,32 +10764,40 @@ static void f_getcompletion(typval_T *argvars, typval_T *rettv, FunPtr fptr)
...
+
+  if (strcmp((char *)get_tv_string(&argvars[1]), "cmdline") == 0) {
```
Comment (justinmk) [inline src/nvim/eval.c]: What happens if `argvars[1]` is *not* "cmdline"? Shouldn't it raise an error?

File: src/nvim/testdir/test_cmdline.vim (hunk 7 above)
Comment (Shougo) [inline src/nvim/testdir/test_cmdline.vim]: Because, it may contain other candidates.

File: src/nvim/eval.c (hunk 6 above)
Comment (Shougo) [inline src/nvim/eval.c]: > Or is this used by set_one_cmd_context() ?

Yes, it is set by `set_one_cmd_context()`.

File: src/nvim/eval.c (hunk 8 above)
Comment (Shougo) [inline src/nvim/eval.c]: It is checked by:

```c
    xpc.xp_context = cmdcomplete_str_to_type(get_tv_string(&argvars[1]));

    if (xpc.xp_context == EXPAND_NOTHING) {
      EMSG2(_(e_invarg2), argvars[1].vval.v_string);
      return;
    }
```

**Comment by Shougo** [comment]
Current known issue:  User defined completion is not supported...

**Comment by Shougo** [comment]
I have removed the cast and rebased.

File: src/nvim/eval.c (hunk 6 above)
Comment (justinmk) [inline src/nvim/eval.c]: If it's set by set_one_cmd_context() then it's not _used_. Which means this line can move down.

File: runtime/doc/eval.txt [hunk 9]
```diff
@@ -4026,14 +4027,19 @@ getcompletion({pat}, {type} [, {filtered}])		*getcompletion()*
...
 
 		If the optional {filtered} flag is set to 1, then 'wildignore'
 		is applied to filter the results.  Otherwise all the matches
 		are returned. The 'wildignorecase' option always applies.
 
+		If {type} is "cmdline", then the |cmdline-completion| result is
+		returned. >
+                	" It returns :call completion result
+                	echo getcompletion('call ', 'cmdline')
+<
```
Comment (justinmk) [inline runtime/doc/eval.txt]: why does this deserve an example? None of the other items have paragraphs like this.

File: src/nvim/eval.c [hunk 10]
```diff
@@ -10765,32 +10765,40 @@ static void f_getcompletion(typval_T *argvars, typval_T *rettv, FunPtr fptr)
     options |= WILD_KEEP_ALL;
   }
 
+  if (argvars[0].v_type != VAR_STRING || argvars[1].v_type != VAR_STRING) {
+    EMSG(_(e_invarg));
+    return;
+  }
+
   ExpandInit(&xpc);
   xpc.xp_pattern = get_tv_string(&argvars[0]);
```
Comment (justinmk) [inline src/nvim/eval.c]: this is already done by `set_one_cmd_context`, so this is a redundant call to `get_tv_string`.

**Comment by justinmk** [comment]
Why wasn't "cmdline" added to `command_complete`?

The reason this wasn't merged quickly is because it doesn't look like the changes were done carefully.

**Comment by justinmk** [comment]
Merged in https://github.com/neovim/neovim/pull/6376, Bram is going to change it anyway. Still wondering whether `command_complete` should have a "cmdline" item added.

**Comment by Shougo** [comment]
> Merged in #6376, Bram is going to change it anyway. Still wondering whether command_complete should have a "cmdline" item added.

Hm.  Bram may change the implementation.

**Pull request closed without merging**